## **Proyecto 1 - MLP para Precios de Casas**

**Bryan Alberto Martínez Orellana**


In [1]:
import pandas as pd

## **Análisis Exploratorio**

### **Descripción general y estructura del dataset**

In [2]:
df = pd.read_csv(
    "data/train.csv",
    keep_default_na=False,
    na_values=[""],
)

cantidad_observaciones, cantidad_variables = df.shape

print(f"Cantidad de observaciones: {cantidad_observaciones}")
print(f"Cantidad de variables: {cantidad_variables}")

display(df.head())

Cantidad de observaciones: 1168
Cantidad de variables: 81


,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
0,255,20,RL,70.0,8400,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,6,2010,WD,Normal,145000
1,1067,60,RL,59.0,7837,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,5,2009,WD,Normal,178000
2,639,30,RL,67.0,8777,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,MnPrv,NaN,0,5,2008,WD,Normal,85000
3,800,50,RL,60.0,7200,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,MnPrv,NaN,0,6,2007,WD,Normal,175000
4,381,50,RL,50.0,5000,Pave,Pave,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,5,2010,WD,Normal,127000


El dataset cuenta con 1168 observaciones y 81 variables, de las cuales 80 son variables predictoras y 1 es la objetivo. 

**Variables Predictoras:**

- `Id`: variable numérica discreta que identifica cada observación del dataset.
- `MSSubClass`: variable categórica nominal que representa la clase de vivienda.
- `MSZoning`: variable categórica nominal que identifica la clasificación de la zona.
- `LotFrontage`: variable numérica continua que mide la longitud del frente del terreno conectado a la calle.
- `LotArea`: variable numérica continua que representa el área total del terreno.
- `Street`: variable categórica nominal que indica el tipo de acceso vial a la propiedad.
- `Alley`: variable categórica nominal que indica el tipo de acceso por callejón.
- `LotShape`: variable categórica ordinal que describe la forma general del terreno.
- `LandContour`: variable categórica nominal que describe el nivel de planitud del terreno.
- `Utilities`: variable categórica nominal que indica los servicios disponibles en la propiedad.
- `LotConfig`: variable categórica nominal que describe la configuración del terreno respecto a las calles.
- `LandSlope`: variable categórica ordinal que representa la inclinación del terreno.
- `Neighborhood`: variable categórica nominal que identifica el vecindario de la propiedad.
- `Condition1`: variable categórica nominal que representa la condición principal cercana a la propiedad.
- `Condition2`: variable categórica nominal que representa una segunda condición cercana a la propiedad.
- `BldgType`: variable categórica nominal que identifica el tipo de vivienda.
- `HouseStyle`: variable categórica nominal que describe el estilo y la cantidad de niveles de la vivienda.
- `OverallQual`: variable categórica ordinal que califica la calidad general de los materiales y acabados.
- `OverallCond`: variable categórica ordinal que califica la condición general de la vivienda.
- `YearBuilt`: variable numérica discreta que indica el año de construcción original.
- `YearRemodAdd`: variable numérica discreta que indica el año de la última remodelación o ampliación.
- `RoofStyle`: variable categórica nominal que identifica el estilo del techo.
- `RoofMatl`: variable categórica nominal que identifica el material del techo.
- `Exterior1st`: variable categórica nominal que representa el material exterior principal de la vivienda.
- `Exterior2nd`: variable categórica nominal que representa un segundo material exterior de la vivienda.
- `MasVnrType`: variable categórica nominal que identifica el tipo de revestimiento de mampostería.
- `MasVnrArea`: variable numérica continua que mide el área del revestimiento de mampostería.
- `ExterQual`: variable categórica ordinal que califica la calidad de los materiales exteriores.
- `ExterCond`: variable categórica ordinal que califica la condición actual de los materiales exteriores.
- `Foundation`: variable categórica nominal que identifica el tipo de cimiento de la vivienda.
- `BsmtQual`: variable categórica ordinal que califica la altura del sótano.
- `BsmtCond`: variable categórica ordinal que califica la condición general del sótano.
- `BsmtExposure`: variable categórica ordinal que indica el nivel de exposición del sótano al exterior.
- `BsmtFinType1`: variable categórica nominal que identifica la calidad del acabado principal del sótano.
- `BsmtFinSF1`: variable numérica continua que mide el área terminada principal del sótano.
- `BsmtFinType2`: variable categórica nominal que identifica la calidad de un segundo acabado del sótano.
- `BsmtFinSF2`: variable numérica continua que mide el área correspondiente al segundo acabado del sótano.
- `BsmtUnfSF`: variable numérica continua que mide el área sin terminar del sótano.
- `TotalBsmtSF`: variable numérica continua que mide el área total del sótano.
- `Heating`: variable categórica nominal que identifica el tipo de calefacción de la vivienda.
- `HeatingQC`: variable categórica ordinal que califica la calidad y condición de la calefacción.
- `CentralAir`: variable categórica nominal que indica si la vivienda cuenta con aire acondicionado central.
- `Electrical`: variable categórica nominal que identifica el tipo de sistema eléctrico.
- `1stFlrSF`: variable numérica continua que mide el área del primer nivel.
- `2ndFlrSF`: variable numérica continua que mide el área del segundo nivel.
- `LowQualFinSF`: variable numérica continua que mide el área terminada de baja calidad.
- `GrLivArea`: variable numérica continua que mide el área habitable ubicada sobre el nivel del suelo.
- `BsmtFullBath`: variable numérica discreta que indica la cantidad de baños completos en el sótano.
- `BsmtHalfBath`: variable numérica discreta que indica la cantidad de medios baños en el sótano.
- `FullBath`: variable numérica discreta que indica la cantidad de baños completos sobre el nivel del suelo.
- `HalfBath`: variable numérica discreta que indica la cantidad de medios baños sobre el nivel del suelo.
- `BedroomAbvGr`: variable numérica discreta que indica la cantidad de dormitorios sobre el nivel del suelo.
- `KitchenAbvGr`: variable numérica discreta que indica la cantidad de cocinas sobre el nivel del suelo.
- `KitchenQual`: variable categórica ordinal que califica la calidad de la cocina.
- `TotRmsAbvGrd`: variable numérica discreta que indica la cantidad total de habitaciones sobre el nivel del suelo.
- `Functional`: variable categórica ordinal que califica la funcionalidad general de la vivienda.
- `Fireplaces`: variable numérica discreta que indica la cantidad de chimeneas.
- `FireplaceQu`: variable categórica ordinal que califica la calidad de la chimenea.
- `GarageType`: variable categórica nominal que identifica el tipo de garaje.
- `GarageYrBlt`: variable numérica discreta que indica el año de construcción del garaje.
- `GarageFinish`: variable categórica ordinal que describe el acabado interior del garaje.
- `GarageCars`: variable numérica discreta que indica la capacidad del garaje en cantidad de vehículos.
- `GarageArea`: variable numérica continua que mide el área del garaje.
- `GarageQual`: variable categórica ordinal que califica la calidad del garaje.
- `GarageCond`: variable categórica ordinal que califica la condición del garaje.
- `PavedDrive`: variable categórica ordinal que indica el nivel de pavimentación de la entrada vehicular.
- `WoodDeckSF`: variable numérica continua que mide el área de la terraza de madera.
- `OpenPorchSF`: variable numérica continua que mide el área del porche abierto.
- `EnclosedPorch`: variable numérica continua que mide el área del porche cerrado.
- `3SsnPorch`: variable numérica continua que mide el área del porche de tres estaciones.
- `ScreenPorch`: variable numérica continua que mide el área del porche con mosquitero.
- `PoolArea`: variable numérica continua que mide el área de la piscina.
- `PoolQC`: variable categórica ordinal que califica la calidad de la piscina.
- `Fence`: variable categórica ordinal que califica la calidad de la cerca.
- `MiscFeature`: variable categórica nominal que identifica una característica adicional no incluida en otras variables.
- `MiscVal`: variable numérica continua que representa el valor monetario de la característica adicional.
- `MoSold`: variable numérica discreta que indica el mes de la venta.
- `YrSold`: variable numérica discreta que indica el año de la venta.
- `SaleType`: variable categórica nominal que identifica el tipo de venta.
- `SaleCondition`: variable categórica nominal que describe la condición en la que se realizó la venta.

**Variable Objetivo:**

- `SalePrice`: variable numérica continua que representa el precio de venta de la vivienda.

### **Estadísticas Descriptivas**

In [3]:
# Estadísticas descriptivas de las variables numéricas
variables_numericas = [
    "SalePrice",
    "LotArea",
    "GrLivArea",
    "TotalBsmtSF",
    "GarageCars",
    "YearBuilt",
    "YearRemodAdd",
    "FullBath",
    "Fireplaces",
]

estadisticas_numericas = (
    df[variables_numericas]
    .describe()
    .T
    .rename(
        columns={
            "count": "Cantidad",
            "mean": "Media",
            "std": "Desviación estándar",
            "min": "Mínimo",
            "25%": "Q1 (25%)",
            "50%": "Mediana",
            "75%": "Q3 (75%)",
            "max": "Máximo",
        }
    )
)

estadisticas_numericas["Rango"] = (
    estadisticas_numericas["Máximo"] - estadisticas_numericas["Mínimo"]
)

estadisticas_numericas = estadisticas_numericas[
    [
        "Cantidad",
        "Media",
        "Mediana",
        "Desviación estándar",
        "Mínimo",
        "Q1 (25%)",
        "Q3 (75%)",
        "Máximo",
        "Rango",
    ]
]

display(estadisticas_numericas.round(2))

# Resumen de la variable ordinal OverallQual
resumen_overall_qual = pd.DataFrame(
    {
        "Mediana": [df["OverallQual"].median()],
        "Moda": [df["OverallQual"].mode().iloc[0]],
    },
    index=["OverallQual"],
)

frecuencias_overall_qual = (
    df["OverallQual"]
    .value_counts(dropna=False)
    .sort_index()
    .rename_axis("Categoría")
    .reset_index(name="Frecuencia")
)
frecuencias_overall_qual["Porcentaje"] = (
    frecuencias_overall_qual["Frecuencia"] / len(df) * 100
).round(2)

print("Resumen de OverallQual")
display(resumen_overall_qual)
display(frecuencias_overall_qual)


# Tablas de frecuencias de las variables categóricas seleccionadas
variables_categoricas = [
    "Neighborhood",
    "KitchenQual",
    "BsmtQual",
    "ExterQual",
    "GarageFinish",
]

orden_categorias = {
    "KitchenQual": ["Ex", "Gd", "TA", "Fa", "Po", "Faltante"],
    "BsmtQual": ["Ex", "Gd", "TA", "Fa", "Po", "Faltante"],
    "ExterQual": ["Ex", "Gd", "TA", "Fa", "Po", "Faltante"],
    "GarageFinish": ["Fin", "RFn", "Unf", "Faltante"],
}

def crear_tabla_frecuencias(variable):
    serie = df[variable].fillna("Faltante")
    frecuencias = serie.value_counts()

    if variable in orden_categorias:
        categorias = [
            categoria
            for categoria in orden_categorias[variable]
            if categoria in frecuencias.index
        ]
        frecuencias = frecuencias.reindex(categorias)

    tabla = frecuencias.rename_axis("Categoría").reset_index(name="Frecuencia")
    tabla["Porcentaje"] = (tabla["Frecuencia"] / len(df) * 100).round(2)

    return tabla

for variable in variables_categoricas:
    print(f"Tabla de frecuencias: {variable}")
    display(crear_tabla_frecuencias(variable))

,Cantidad,Media,Mediana,Desviación estándar,Mínimo,Q1 (25%),Q3 (75%),Máximo,Rango
SalePrice,1168.0,181441.54,165000.0,77263.58,34900.0,130000.00,214925.00,745000.0,710100.0
LotArea,1168.0,10689.64,9600.0,10759.37,1300.0,7587.25,11700.00,215245.0,213945.0
GrLivArea,1168.0,1527.40,1473.0,524.43,334.0,1145.75,1792.00,5642.0,5308.0
TotalBsmtSF,1168.0,1061.77,997.5,440.68,0.0,796.00,1299.25,6110.0,6110.0
GarageCars,1168.0,1.78,2.0,0.74,0.0,1.00,2.00,4.0,4.0
YearBuilt,1168.0,1970.97,1972.0,30.68,1872.0,1953.00,2001.00,2010.0,138.0
YearRemodAdd,1168.0,1984.90,1994.0,20.73,1950.0,1966.00,2004.00,2010.0,60.0
FullBath,1168.0,1.58,2.0,0.55,0.0,1.00,2.00,3.0,3.0
Fireplaces,1168.0,0.62,1.0,0.65,0.0,0.00,1.00,3.0,3.0


Resumen de OverallQual


,Mediana,Moda
OverallQual,6.0,6


,Categoría,Frecuencia,Porcentaje
0,1,2,0.17
1,2,1,0.09
2,3,16,1.37
3,4,88,7.53
4,5,308,26.37
5,6,311,26.63
6,7,258,22.09
7,8,135,11.56
8,9,36,3.08
9,10,13,1.11


Tabla de frecuencias: Neighborhood


,Categoría,Frecuencia,Porcentaje
0,NAmes,181,15.50
1,CollgCr,115,9.85
2,OldTown,91,7.79
3,Edwards,87,7.45
4,Somerst,69,5.91
5,NWAmes,66,5.65
6,Gilbert,65,5.57
7,NridgHt,61,5.22
8,Sawyer,58,4.97
9,BrkSide,45,3.85


Tabla de frecuencias: KitchenQual


,Categoría,Frecuencia,Porcentaje
0,Ex,77,6.59
1,Gd,470,40.24
2,TA,589,50.43
3,Fa,32,2.74


Tabla de frecuencias: BsmtQual


,Categoría,Frecuencia,Porcentaje
0,Ex,97,8.30
1,Gd,493,42.21
2,TA,521,44.61
3,Fa,29,2.48
4,Faltante,28,2.40


Tabla de frecuencias: ExterQual


,Categoría,Frecuencia,Porcentaje
0,Ex,42,3.60
1,Gd,388,33.22
2,TA,727,62.24
3,Fa,11,0.94


Tabla de frecuencias: GarageFinish


,Categoría,Frecuencia,Porcentaje
0,Fin,285,24.40
1,RFn,339,29.02
2,Unf,480,41.10
3,Faltante,64,5.48


#### **Discusión de las estadísticas descriptivas**

Primero, al analizar las variables numéricas seleccionadas se observa que todas cuentan con 1168 registros, por lo que no presentan valores faltantes dentro de este subconjunto. En el caso de `SalePrice`, la mediana es de $165,000, es decir, este es el valor que divide las observaciones en dos mitades, mientras que la media aumenta hasta $181,441.54. Esta diferencia sugiere un sesgo hacia la derecha, ya que parece existir un grupo de viviendas con precios elevados que desplaza el promedio por encima de la mediana. Asimismo, la desviación estándar es de $77,263.58, lo que muestra una dispersión considerable en los precios, y esto también se refleja al comparar el mínimo de $34,900 con el máximo de $745,000. El primer cuartil se encuentra en $130,000 y el tercero en $214,925, por lo que el 50 % central de las viviendas se concentra entre estos valores. Sin embargo, entre el tercer cuartil y el máximo existe una diferencia superior a $500,000, además de un rango total de $710,100, lo cual sugiere la posible presencia de observaciones extremas en los precios. No obstante, esta diferencia por sí sola no permite afirmar cuántos valores atípicos existen, de tal forma que posteriormente se deberá confirmar mediante el rango intercuartílico y las visualizaciones correspondientes.

Luego, `LotArea`, que representa el área completa del terreno, presenta una media de 10,689.64 pies cuadrados y una mediana de 9,600 pies cuadrados. Nuevamente, la media se encuentra por encima de la mediana, lo que indica que algunos terrenos de gran tamaño podrían estar desplazando el promedio hacia valores mayores. La desviación estándar es de 10,759.37 pies cuadrados, incluso ligeramente superior a la media, y esto muestra una dispersión bastante elevada. El mínimo es de 1,300 pies cuadrados, mientras que el primer cuartil se encuentra en 7,587.25 y el tercero en 11,700 pies cuadrados; sin embargo, el máximo alcanza 215,245 pies cuadrados, un valor más de 18 veces superior al tercer cuartil. Bajo esta idea, el rango de 213,945 pies cuadrados parece estar determinado en buena medida por un grupo reducido de terrenos muy grandes, por lo que esta variable también podría presentar un sesgo positivo y valores atípicos.

Asimismo, `GrLivArea` tiene una media de 1,527.40 pies cuadrados y una mediana de 1,473, valores relativamente cercanos entre sí, mientras que su desviación estándar es de 524.43 pies cuadrados. El mínimo corresponde a 334 pies cuadrados, el primer cuartil a 1,145.75 y el tercero a 1,792, de tal forma que la mitad central de las viviendas se encuentra dentro de un intervalo considerablemente más estable que el rango completo. El máximo alcanza 5,642 pies cuadrados y produce un salto importante respecto al tercer cuartil, por lo que considero que también se debe comprobar si las viviendas con áreas habitables muy grandes corresponden a valores atípicos.

En cuanto a `TotalBsmtSF`, se obtiene una media de 1,061.77 pies cuadrados y una mediana de 997.5, con una desviación estándar de 440.68 pies cuadrados. El mínimo es de cero, lo que indica que algunas viviendas no cuentan con sótano, mientras que el primer cuartil se encuentra en 796 y el tercero en 1,299.25 pies cuadrados. Por otro lado, el máximo llega hasta 6,110 pies cuadrados y el rango tiene este mismo valor, dado que el mínimo es cero. Aunque la media y la mediana son relativamente cercanas, la distancia entre el tercer cuartil y el máximo vuelve a mostrar la posible presencia de sótanos con áreas considerablemente mayores al resto.

Después, al analizar `GarageCars`, se observa una media de 1.78 espacios y una mediana de 2, por lo que la capacidad típica del garaje se encuentra alrededor de uno o dos vehículos. La desviación estándar es de 0.74, el mínimo es de cero y el máximo de cuatro espacios. Además, el primer cuartil corresponde a un vehículo y el tercero a dos, lo que indica que el 50 % central de las viviendas tiene capacidad para entre uno y dos automóviles. El valor mínimo también permite identificar viviendas sin espacio de garaje, mientras que los garajes con capacidad para cuatro vehículos se encuentran en el extremo superior de la distribución.

Al revisar `YearBuilt`, la media corresponde aproximadamente a 1971 y la mediana a 1972, por lo que ambas medidas ubican el centro de los datos en un periodo similar. Sin embargo, la desviación estándar es de 30.68 años y el conjunto abarca desde 1872 hasta 2010, produciendo un rango de 138 años. El primer cuartil se encuentra en 1953 y el tercero en 2001, es decir, incluso el 50 % central cubre un periodo de 48 años. Esto muestra que el dataset contiene viviendas de antigüedades bastante diferentes, por lo que la edad de la propiedad podría ser un factor importante para poder explicar el precio.

En `YearRemodAdd`, la media se encuentra aproximadamente en 1985 y la mediana en 1994, mientras que la desviación estándar es de 20.73 años. El valor mínimo corresponde a 1950, el primer cuartil a 1966, el tercero a 2004 y el máximo a 2010, dando como resultado un rango de 60 años. La mediana posterior a la media sugiere que las fechas más antiguas desplazan el promedio hacia abajo. Adicional, aunque esta variable presenta años más recientes que `YearBuilt`, no es tan directo concluir únicamente a partir de estas estadísticas cuántas viviendas fueron remodeladas, ya que cuando una propiedad no tuvo remodelación la variable puede conservar el año de construcción. Para poder establecerlo sería necesario comparar ambas columnas por observación.

Para `FullBath`, la media es de 1.58 baños completos y la mediana es de 2, con una desviación estándar de 0.55. El mínimo es de cero, el primer cuartil de uno, el tercero de dos y el máximo de tres baños, por lo que la mayoría de las viviendas parece concentrarse entre uno y dos baños completos. Por otro lado, `Fireplaces` trabaja con una escala similar pero presenta una media de 0.62 chimeneas y una mediana de 1, con una desviación estándar de 0.65. Tanto el mínimo como el primer cuartil son cero, el tercer cuartil es uno y el máximo es tres, lo que indica que la mitad central de las viviendas tiene entre cero y una chimenea, mientras que únicamente el extremo superior alcanza cantidades mayores.

Por otro lado, `OverallQual` se trató como una variable ordinal, dado que sus valores representan niveles de calidad y no una medida continua. Tanto la mediana como la moda son 6, lo que indica que el centro y la categoría más frecuente corresponden a una calidad ligeramente superior al punto medio de la escala. Esto también se refleja en la tabla de frecuencias, ya que las categorías 5 y 6 reúnen 308 y 311 viviendas, respectivamente, mientras que la categoría 7 contiene 258. Después, la frecuencia disminuye a 135 viviendas en la categoría 8, 36 en la 9 y 13 en la 10. En el extremo inferior únicamente aparecen dos viviendas con calidad 1 y una con calidad 2, por lo que la distribución se encuentra principalmente concentrada en calidades medias y medias-altas, con pocos casos en ambos extremos.

En el caso de `Neighborhood`, se observa una distribución desigual entre las categorías. Por ejemplo, `NAmes` contiene 181 viviendas y representa el 15.50 % de los datos, seguido por `CollgCr` con 115 observaciones y un 9.85 %. A partir de estas categorías la frecuencia disminuye de forma progresiva hasta llegar a `Blueste`, que cuenta con una sola observación y representa el 0.09 %. Esta diferencia implica que algunos vecindarios tienen bastante mayor representación que otros, por lo que posteriormente se deberá considerar cómo tratar las categorías con muy pocos registros para evitar que el modelo aprenda relaciones poco estables.

Finalmente, las variables relacionadas con la calidad también muestran una concentración en categorías intermedias. En `KitchenQual`, la categoría típica o promedio (`TA`) reúne 589 viviendas, equivalentes al 50.43 %, seguida por la categoría buena (`Gd`) con 470 observaciones y un 40.24 %. En `BsmtQual` se mantiene un patrón similar, ya que `TA` representa el 44.61 % y `Gd` el 42.21 %, aunque también aparecen 28 valores faltantes, equivalentes al 2.40 %. En `ExterQual`, la concentración es todavía mayor en `TA`, con 727 observaciones y un 62.24 %, seguida por `Gd` con 388 y un 33.22 %. Por último, en `GarageFinish` predomina la categoría sin terminar (`Unf`) con 480 viviendas y un 41.10 %, seguida por el acabado parcial (`RFn`) con 339 y el acabado completo (`Fin`) con 285, además de 64 valores faltantes. Cabe mencionar que los faltantes de `BsmtQual` y `GarageFinish` podrían representar la ausencia estructural de un sótano o garaje, y no necesariamente información desconocida, por lo que esta distinción deberá validarse antes de decidir el tratamiento de valores faltantes.

### **Identificación y tratamiento de valores faltantes, atípicos e inconsistencias**

Para esta parte se busca diferenciar entre un valor realmente desconocido y una ausencia que describe la vivienda, ya que reemplazar ambos casos de la misma forma podría eliminar información útil para el modelo. Asimismo, los valores atípicos se utilizarán primero como señales para revisar observaciones y no como una regla automática de eliminación, de tal forma que únicamente se descarte un registro cuando exista una contradicción que no pueda explicarse o recuperarse.

In [ ]:
# Resumen de valores faltantes reales
resumen_faltantes = pd.DataFrame(
    {
        "Cantidad": df.isna().sum(),
        "Porcentaje": (df.isna().mean() * 100).round(2),
    }
)
resumen_faltantes = (
    resumen_faltantes[resumen_faltantes["Cantidad"] > 0]
    .sort_values("Porcentaje", ascending=False)
)

display(resumen_faltantes)

# Comprobación de ausencias estructurales
comprobaciones_ausencia = pd.DataFrame(
    [
        {
            "Grupo": "Sótano",
            "Faltantes": df["BsmtQual"].isna().sum(),
            "Coinciden con ausencia": (
                df["BsmtQual"].isna() & df["TotalBsmtSF"].eq(0)
            ).sum(),
        },
        {
            "Grupo": "Chimenea",
            "Faltantes": df["FireplaceQu"].isna().sum(),
            "Coinciden con ausencia": (
                df["FireplaceQu"].isna() & df["Fireplaces"].eq(0)
            ).sum(),
        },
        {
            "Grupo": "Garaje",
            "Faltantes": df["GarageType"].isna().sum(),
            "Coinciden con ausencia": (
                df["GarageType"].isna() & df["GarageCars"].eq(0)
            ).sum(),
        },
        {
            "Grupo": "Piscina",
            "Faltantes": df["PoolQC"].isna().sum(),
            "Coinciden con ausencia": (
                df["PoolQC"].isna() & df["PoolArea"].eq(0)
            ).sum(),
        },
        {
            "Grupo": "Característica adicional",
            "Faltantes": df["MiscFeature"].isna().sum(),
            "Coinciden con ausencia": (
                df["MiscFeature"].isna() & df["MiscVal"].eq(0)
            ).sum(),
        },
    ]
)

display(comprobaciones_ausencia)

# Los seis faltantes reales relacionados con la mampostería
columnas_revision_mamposteria = [
    "Id",
    "MasVnrType",
    "MasVnrArea",
    "Neighborhood",
    "OverallQual",
    "YearBuilt",
    "GrLivArea",
    "SalePrice",
]
casos_mamposteria = df.loc[
    df["MasVnrType"].isna() | df["MasVnrArea"].isna(),
    columnas_revision_mamposteria,
]

display(casos_mamposteria)

#### **Criterio para el tratamiento de valores faltantes**

Primero, se identificaron las variables con observaciones vacías y se compararon con otras características relacionadas para poder establecer si la ausencia tenía un significado propio. Los faltantes del sótano, chimenea, garaje, piscina y características adicionales coinciden con áreas o cantidades iguales a cero, por lo que considero que deben representarse mediante categorías explícitas como `SinSotano`, `SinChimenea` o `SinGaraje`, ya que reemplazarlos con la moda haría que una vivienda sin esa característica pareciera tenerla.

En el caso de `LotFrontage`, los 217 valores faltantes no representan necesariamente que la vivienda no tenga frente hacia la calle, por lo que se utilizará la mediana del vecindario y se conservará un indicador que permita identificar cuáles valores fueron imputados. Para `Electrical`, que contiene un solo faltante, se utilizará la moda. Por otro lado, los seis faltantes simultáneos de `MasVnrType` y `MasVnrArea` corresponden a viviendas coherentes con el resto del dataset, de tal forma que se conservarán las observaciones, se asignará `Desconocido` al tipo y cero al área. No se agregará un indicador adicional para este caso, ya que la categoría `Desconocido` permite reconocer las mismas seis filas.

Cabe mencionar que este tratamiento se aplica sobre una copia para poder documentar el EDA. Cuando se construya el modelo, las medianas y modas deberán calcularse únicamente con los datos de entrenamiento de cada partición para evitar fuga de información.

In [7]:
# Copia tratada para continuar con el EDA sin alterar los datos originales
df_tratado = df.copy()

# Normalización de etiquetas categóricas
columnas_texto = df_tratado.select_dtypes(include="object").columns
for columna in columnas_texto:
    df_tratado[columna] = (
        df_tratado[columna]
        .str.strip()
        .str.strip("'")
    )

# Ausencias estructurales
rellenos_estructurales = {
    "Alley": "SinCallejon",
    "BsmtQual": "SinSotano",
    "BsmtCond": "SinSotano",
    "BsmtExposure": "SinSotano",
    "BsmtFinType1": "SinSotano",
    "BsmtFinType2": "SinSotano",
    "FireplaceQu": "SinChimenea",
    "GarageType": "SinGaraje",
    "GarageFinish": "SinGaraje",
    "GarageQual": "SinGaraje",
    "GarageCond": "SinGaraje",
    "PoolQC": "SinPiscina",
    "Fence": "SinCerca",
    "MiscFeature": "SinCaracteristica",
}

for columna, categoria in rellenos_estructurales.items():
    df_tratado[columna] = df_tratado[columna].fillna(categoria)

# LotFrontage: mediana por vecindario e indicador de imputación
df_tratado["LotFrontageMissing"] = df_tratado["LotFrontage"].isna().astype(int)
mediana_frente_vecindario = df_tratado.groupby("Neighborhood")["LotFrontage"].transform("median")
df_tratado["LotFrontage"] = (
    df_tratado["LotFrontage"]
    .fillna(mediana_frente_vecindario)
    .fillna(df_tratado["LotFrontage"].median())
)

# Mampostería: conservar las observaciones sin inventar un tipo
df_tratado["MasVnrType"] = df_tratado["MasVnrType"].fillna("Desconocido")
df_tratado["MasVnrArea"] = df_tratado["MasVnrArea"].fillna(0)

# Electrical: imputación mediante la categoría más frecuente
moda_electrical = df_tratado["Electrical"].mode().iloc[0]
df_tratado["Electrical"] = df_tratado["Electrical"].fillna(moda_electrical)

# GarageYrBlt: representar la ausencia sin inventar un año
df_tratado["HasGarage"] = df_tratado["GarageCars"].gt(0).astype(int)
df_tratado["GarageAge"] = (
    df_tratado["YrSold"] - df_tratado["GarageYrBlt"]
).where(df_tratado["HasGarage"].eq(1), 0)
df_tratado = df_tratado.drop(columns="GarageYrBlt")

print(f"Valores faltantes restantes: {df_tratado.isna().sum().sum()}")

Valores faltantes restantes: 0


#### **Auditoría de valores atípicos**

La auditoría se realizará sobre todas las variables, aunque el método dependerá de su naturaleza. Para las medidas continuas se utilizará el rango intercuartílico como una primera señal, mientras que las variables con una gran cantidad de ceros se analizarán únicamente entre las viviendas donde la característica está presente. En los conteos y variables ordinales se validarán los rangos y frecuencias, y para los años se revisará la coherencia temporal. Bajo esta idea, una observación marcada no se eliminará automáticamente, sino que pasará a una revisión contextual.

In [8]:
# Variables continuas que serán auditadas mediante IQR
variables_continuas = [
    "LotFrontage", "LotArea", "MasVnrArea", "BsmtFinSF1",
    "BsmtFinSF2", "BsmtUnfSF", "TotalBsmtSF", "1stFlrSF",
    "2ndFlrSF", "LowQualFinSF", "GrLivArea", "GarageArea",
    "WoodDeckSF", "OpenPorchSF", "EnclosedPorch",
    "3SsnPorch", "ScreenPorch", "PoolArea", "MiscVal",
    "SalePrice",
]

variables_con_exceso_ceros = {
    "MasVnrArea", "BsmtFinSF2", "LowQualFinSF", "WoodDeckSF",
    "OpenPorchSF", "EnclosedPorch", "3SsnPorch",
    "ScreenPorch", "PoolArea", "MiscVal",
}

def resumir_outliers_iqr(datos, variable):
    serie_original = datos[variable].dropna()
    serie_evaluada = (
        serie_original[serie_original > 0]
        if variable in variables_con_exceso_ceros
        else serie_original
    )

    q1 = serie_evaluada.quantile(0.25)
    q3 = serie_evaluada.quantile(0.75)
    iqr = q3 - q1
    limite_inferior = q1 - 1.5 * iqr
    limite_superior = q3 + 1.5 * iqr
    mascara = (serie_evaluada < limite_inferior) | (serie_evaluada > limite_superior)

    return {
        "Variable": variable,
        "Observaciones evaluadas": len(serie_evaluada),
        "Ceros excluidos": int(serie_original.eq(0).sum()),
        "Límite inferior": limite_inferior,
        "Límite superior": limite_superior,
        "Candidatos": int(mascara.sum()),
        "Porcentaje evaluado": round(mascara.mean() * 100, 2),
    }

resumen_outliers = pd.DataFrame(
    [resumir_outliers_iqr(df, variable) for variable in variables_continuas]
).sort_values("Porcentaje evaluado", ascending=False)

display(resumen_outliers.round(2))

# Rangos y cardinalidad de variables discretas u ordinales
variables_discretas = [
    "MSSubClass", "OverallQual", "OverallCond", "BsmtFullBath",
    "BsmtHalfBath", "FullBath", "HalfBath", "BedroomAbvGr",
    "KitchenAbvGr", "TotRmsAbvGrd", "Fireplaces", "GarageCars",
    "MoSold",
]
resumen_discretas = df[variables_discretas].agg(["min", "max", "nunique"]).T
resumen_discretas.columns = ["Mínimo", "Máximo", "Valores únicos"]
display(resumen_discretas)

# Categorías poco frecuentes: se señalan, pero no se tratan como errores
resumen_categoricas = []
for variable in df.select_dtypes(include="object").columns:
    frecuencias = df[variable].fillna("Faltante").value_counts()
    resumen_categoricas.append(
        {
            "Variable": variable,
            "Categorías": len(frecuencias),
            "Frecuencia mínima": frecuencias.min(),
            "Categorías con 5 casos o menos": int((frecuencias <= 5).sum()),
        }
    )

display(
    pd.DataFrame(resumen_categoricas)
    .sort_values("Categorías con 5 casos o menos", ascending=False)
)

,Variable,Observaciones evaluadas,Ceros excluidos,Límite inferior,Límite superior,Candidatos,Porcentaje evaluado
15,3SsnPorch,20,1148,15.00,401.00,2,10.00
18,MiscVal,44,1124,-675.00,2325.00,4,9.09
0,LotFrontage,951,0,27.50,111.50,68,7.15
13,OpenPorchSF,656,512,-71.62,225.38,39,5.95
16,ScreenPorch,97,1071,40.00,312.00,5,5.15
12,WoodDeckSF,550,618,-56.88,434.12,27,4.91
2,MasVnrArea,488,674,-211.50,640.50,23,4.71
1,LotArea,1168,0,1418.12,17869.12,54,4.62
19,SalePrice,1168,0,2612.50,342312.50,48,4.11
6,TotalBsmtSF,1168,28,41.12,2054.12,47,4.02


,Mínimo,Máximo,Valores únicos
MSSubClass,20,190,15
OverallQual,1,10,10
OverallCond,1,9,9
BsmtFullBath,0,3,4
BsmtHalfBath,0,2,3
FullBath,0,3,4
HalfBath,0,2,3
BedroomAbvGr,0,8,8
KitchenAbvGr,0,3,4
TotRmsAbvGrd,2,14,12


,Variable,Categorías,Frecuencia mínima,Categorías con 5 casos o menos
10,Condition2,8,1,7
41,SaleType,9,2,5
15,Exterior1st,15,1,5
14,RoofMatl,7,1,5
16,Exterior2nd,16,1,4
40,MiscFeature,5,1,3
38,PoolQC,4,2,3
26,Heating,6,1,3
20,Foundation,6,3,2
13,RoofStyle,6,2,2


In [9]:
# Controles de consistencia lógica
controles_consistencia = pd.DataFrame(
    {
        "Control": [
            "Identificadores duplicados",
            "Filas duplicadas",
            "YearBuilt posterior a YrSold",
            "YearRemodAdd anterior a YearBuilt",
            "YearRemodAdd posterior a YrSold",
            "GarageYrBlt posterior a YrSold",
            "GarageYrBlt anterior a YearBuilt",
            "MoSold fuera del rango 1 a 12",
            "SalePrice menor o igual a cero",
            "GrLivArea no coincide con sus componentes",
            "TotalBsmtSF no coincide con sus componentes",
        ],
        "Cantidad": [
            df["Id"].duplicated().sum(),
            df.duplicated().sum(),
            (df["YearBuilt"] > df["YrSold"]).sum(),
            (df["YearRemodAdd"] < df["YearBuilt"]).sum(),
            (df["YearRemodAdd"] > df["YrSold"]).sum(),
            (df["GarageYrBlt"] > df["YrSold"]).sum(),
            (df["GarageYrBlt"] < df["YearBuilt"]).sum(),
            (~df["MoSold"].between(1, 12)).sum(),
            (df["SalePrice"] <= 0).sum(),
            (
                df["GrLivArea"]
                != df["1stFlrSF"] + df["2ndFlrSF"] + df["LowQualFinSF"]
            ).sum(),
            (
                df["TotalBsmtSF"]
                != df["BsmtFinSF1"] + df["BsmtFinSF2"] + df["BsmtUnfSF"]
            ).sum(),
        ],
    }
)

display(controles_consistencia)

# Filas que requieren revisión temporal, sin eliminarlas
mascara_revision_temporal = (
    (df["YearBuilt"] > df["YrSold"])
    | (df["YearRemodAdd"] < df["YearBuilt"])
    | (df["YearRemodAdd"] > df["YrSold"])
    | (df["GarageYrBlt"] > df["YrSold"])
    | (df["GarageYrBlt"] < df["YearBuilt"])
)
columnas_temporales = [
    "Id", "YearBuilt", "YearRemodAdd", "GarageYrBlt",
    "YrSold", "SaleType", "SaleCondition", "SalePrice",
]
display(df.loc[mascara_revision_temporal, columnas_temporales])

# Etiquetas con comillas externas que deben normalizarse
etiquetas_con_formato = []
for variable in df.select_dtypes(include="object").columns:
    valores = [
        valor
        for valor in df[variable].dropna().unique()
        if str(valor).startswith("'") or str(valor).endswith("'")
    ]
    if valores:
        etiquetas_con_formato.append(
            {"Variable": variable, "Etiquetas": valores}
        )

display(pd.DataFrame(etiquetas_con_formato))

,Control,Cantidad
0,Identificadores duplicados,0
1,Filas duplicadas,0
2,YearBuilt posterior a YrSold,0
3,YearRemodAdd anterior a YearBuilt,0
4,YearRemodAdd posterior a YrSold,1
5,GarageYrBlt posterior a YrSold,0
6,GarageYrBlt anterior a YearBuilt,5
7,MoSold fuera del rango 1 a 12,0
8,SalePrice menor o igual a cero,0
9,GrLivArea no coincide con sus componentes,0


,Id,YearBuilt,YearRemodAdd,GarageYrBlt,YrSold,SaleType,SaleCondition,SalePrice
72,1419,1963,1963,1962.0,2008,COD,Normal,124000
231,1377,1930,1950,1925.0,2008,WD,Normal,91000
288,1104,1959,1959,1954.0,2006,WD,Normal,159500
365,524,2007,2008,2007.0,2007,New,Partial,184750
769,94,1910,1998,1900.0,2007,WD,Normal,133900
1117,601,2005,2005,2003.0,2006,WD,Normal,275000


,Variable,Etiquetas
0,MSZoning,['C (all)']
1,Exterior1st,['Wd Sdng']
2,Exterior2nd,"['Wd Sdng', 'Wd Shng', 'Brk Cmn']"


In [10]:
# Revisión contextual de los valores atípicos de SalePrice
q1_precio = df["SalePrice"].quantile(0.25)
q3_precio = df["SalePrice"].quantile(0.75)
iqr_precio = q3_precio - q1_precio
limite_inferior_precio = q1_precio - 1.5 * iqr_precio
limite_superior_precio = q3_precio + 1.5 * iqr_precio

mascara_precio_atipico = (
    (df["SalePrice"] < limite_inferior_precio)
    | (df["SalePrice"] > limite_superior_precio)
)

casos_precio_atipico = df.loc[
    mascara_precio_atipico,
    [
        "Id", "SalePrice", "GrLivArea", "LotArea",
        "OverallQual", "OverallCond", "YearBuilt",
        "Neighborhood", "SaleType", "SaleCondition",
    ],
].copy()
casos_precio_atipico["PrecioPorPieCuadrado"] = (
    casos_precio_atipico["SalePrice"] / casos_precio_atipico["GrLivArea"]
).round(2)

print(f"Casos señalados por el IQR de SalePrice: {len(casos_precio_atipico)}")
display(casos_precio_atipico.sort_values("SalePrice", ascending=False))

# Comparación contextual del 1 % de viviendas con mayor área habitable
limite_area_extrema = df["GrLivArea"].quantile(0.99)
casos_area_extrema = df.loc[
    df["GrLivArea"] >= limite_area_extrema,
    [
        "Id", "SalePrice", "GrLivArea", "LotArea",
        "OverallQual", "OverallCond", "YearBuilt",
        "Neighborhood", "SaleType", "SaleCondition",
    ],
].copy()
casos_area_extrema["PrecioPorPieCuadrado"] = (
    casos_area_extrema["SalePrice"] / casos_area_extrema["GrLivArea"]
).round(2)

display(casos_area_extrema.sort_values("PrecioPorPieCuadrado"))

Casos señalados por el IQR de SalePrice: 48


,Id,SalePrice,GrLivArea,LotArea,OverallQual,OverallCond,YearBuilt,Neighborhood,SaleType,SaleCondition,PrecioPorPieCuadrado
173,1183,745000,4476,15623,10,5,1996,NoRidge,WD,Abnorml,166.44
539,1170,625000,3627,35760,10,5,1995,NoRidge,WD,Normal,172.32
593,804,582933,2822,13891,9,5,2008,NridgHt,New,Partial,206.57
537,441,555000,2402,15431,10,5,2008,NridgHt,WD,Normal,231.06
1151,770,538000,3279,53504,8,5,2003,StoneBr,WD,Normal,164.07
208,179,501837,2234,17423,9,5,2008,StoneBr,New,Partial,224.64
145,799,485000,3140,13518,9,5,2008,NridgHt,New,Partial,154.46
638,186,475000,3608,22950,10,9,1892,OldTown,WD,Normal,131.65
882,1374,466500,2633,11400,10,5,2001,NoRidge,WD,Normal,177.17
103,528,446261,2713,14948,9,5,2008,NridgHt,New,Partial,164.49


,Id,SalePrice,GrLivArea,LotArea,OverallQual,OverallCond,YearBuilt,Neighborhood,SaleType,SaleCondition,PrecioPorPieCuadrado
495,1299,160000,5642,63887,10,5,2008,Edwards,New,Partial,28.36
365,524,184750,4676,40094,10,5,2007,Edwards,New,Partial,39.51
792,636,200000,3395,10896,6,7,1914,SWISU,WD,Abnorml,58.91
815,305,295000,3493,18386,7,9,1880,OldTown,WD,Normal,84.45
482,119,320000,3222,12376,7,5,1990,SawyerW,WD,Normal,99.32
806,1269,381000,3447,14100,8,9,1935,Crawfor,WD,Normal,110.53
290,1354,410000,3238,14720,8,5,1995,NoRidge,WD,Normal,126.62
638,186,475000,3608,22950,10,9,1892,OldTown,WD,Normal,131.65
866,497,430000,3228,12692,8,5,1992,NoRidge,WD,Normal,133.21
1151,770,538000,3279,53504,8,5,2003,StoneBr,WD,Normal,164.07


#### **Criterio para la revisión de outliers e inconsistencias**

Los límites calculados permiten reducir la cantidad de observaciones que se deben revisar, pero no demuestran que una vivienda sea incorrecta. Por ejemplo, un precio elevado podría explicarse mediante una calidad alta, un vecindario con mayor valor o una construcción reciente. En cambio, los casos donde el área, la calidad, el precio y las fechas parecen contradecirse requieren una revisión más profunda, como sucede inicialmente con las observaciones `Id = 524` e `Id = 1299`.

Dicho esto, se conservarán los casos raros que mantengan coherencia entre sus variables. Si una observación continúa sin poder explicarse, se comparará el mismo modelo con y sin ella utilizando un conjunto de validación fijo y varias semillas. Solamente se considerará su eliminación cuando exista evidencia de inconsistencia y la mejora del RMSE sea estable, buscando evitar que una decisión tomada a partir de un solo resultado termine reduciendo la capacidad de generalización.